crux-core/src/crux/api/transaction/Transaction.java

```diff
@@ -0,0 +1,122 @@
+package crux.api.transaction;
+
+import clojure.lang.PersistentVector;
+import crux.api.IBuilder;
+import crux.api.document.ICruxDocument;
+import crux.api.transaction.operation.*;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class Transaction {
```

Comment:
Thinking on this - would it be easier to specify the Builder as an interface that we then implement in Clojure? From a Java user's PoV it'd look similar, but we don't have to worry about building up an object tree in Java that we then translate back into Clojure?

Reply:
This might be a crazy idea, it's been a long night :baby:, but

Java side doesn't differ hugely from what you have:

```java
public interface ITransactionBuilder {
    ITransactionBuilder put(ICruxDocument doc);
    ITransactionBuilder put(ICruxDocument doc, Date fromValidTime);
    ITransactionBuilder put(ICruxDocument doc, Date startValidTime, Date endValidTime);
}

interface ICruxAPI {
    Map<Keyword, ?> submitTx(Consumer<ITransactionBuilder> f);
}
```

On the Clojure side:
```clojure
(defn <-i-crux-doc [^ICruxDocument doc]
  (assoc (.getData doc)
         :crux.db/id (.getId doc)))

(defrecord TransactionBuilder [!tx-ops]
  ITransactionBuilder
  (put [this doc]
    (swap! !tx-ops conj [:crux.tx/put (<-i-crux-doc doc)]))
  ...)

(defrecord CruxNode [...]
  ICruxAPI
  (submitTx [this ^Consumer c]
    (let [builder (->TransactionBuilder (atom []))]
      (.accept c builder)
      (.submitTx this @(:!tx-ops builder)))))
```

Tradeoff is that we lose the Transaction as Java data. The documents themselves still will be, the user supplies those, but we lose the tx-events as Java data.

wdyt? worth a spike?